lib/hanami/application.rb

```diff
@@ -1,64 +1,282 @@
     @_mutex = Mutex.new
 
-    def self.inherited(app_class)
-      @_mutex.synchronize do
-        app_class.class_eval do
-          @_mutex         = Mutex.new
-          @_configuration = Hanami::Configuration.new(env: Hanami.env)
+    class << self
+      def inherited(klass)
+        @_mutex.synchronize do
+          klass.class_eval do
+            @_mutex         = Mutex.new
+            @_configuration = Hanami::Configuration.new(env: Hanami.env)
 
-          extend ClassMethods
-          include InstanceMethods
-        end
+            extend ClassMethods
+            include InstanceMethods
+          end
 
-        Hanami.application_class = app_class
+          Hanami.application = klass
+        end
       end
     end
 
-    # Class method interface
-    #
-    # @since 2.0.0
     module ClassMethods
       def configuration
         @_configuration
       end
 
-      alias config configuration
+      alias_method :config, :configuration
+
+      def init
+        return self if inited?
+
+        configuration.finalize
+
+        @container = prepare_container
+        @deps_module = prepare_deps_module
+        @slices = load_slices
+        @inited = true
+        self
+      end
+
+      def container
+        raise "Application not init'ed" unless defined?(@container)
+        @container
+      end
+
+      def deps
+        raise "Application not init'ed" unless defined?(@deps_module)
+        @deps_module
+      end
+
+      def slices
+        raise "Application not init'ed" unless defined?(@slices)
+        @slices
+      end
+
+      def register_bootable(*args, &block)
+        container.boot(*args, &block)
+      end
+
+      def init_bootable(*args)
+        container.init(*args)
+      end
+
+      def start_bootable(*args)
+        container.start(*args)
+      end
+
+      def keys
+        container.keys
+      end
+
+      def [](*args)
+        container.[](*args)
+      end
+
+      def resolve(*args)
+        container.resolve(*args)
+      end
+
+      def boot(&block)
+        return self if booted?
+
+        init
+
+        container.configure do; end # force after configure hook
 
-      def routes(&blk)
+        container.finalize!(&block)
+
+        slices.each do |slice|
+          slice.boot
+        end
+
+        @booted = true
+        self
+      end
+
+      def routes(&block)
         @_mutex.synchronize do
-          if blk.nil?
-            raise "Hanami.application_class.routes not configured" unless defined?(@_routes)
+          if block.nil?
+            raise "Hanami.application.routes not configured" unless defined?(@_routes)
 
             @_routes
           else
-            @_routes = Routes.new(&blk)
+            @_routes = block
           end
         end
       end
+
+      MODULE_DELIMITER = "::"
+      private_constant :MODULE_DELIMITER
+
+      def application_module
+        inflector.constantize(name.split(MODULE_DELIMITER)[0..-2].join(MODULE_DELIMITER))
+      end
```

Comment:
I wonder if `application_namespace` would be a better name. `module` is less clear.

Reply:
I like it! Plus it is more consistent with the `namespace` that we're passing to `Slice` objects